primitives/src/addr.rs

```diff
@@ -6,30 +6,96 @@
 #[cfg(not(feature = "std"))]
 use alloc::{format, string::String, vec::Vec};
 
-use derive_more::{AsRef, From, IntoIterator};
+use derive_more::{AsRef, Display, From, IntoIterator};
+use iroha_macro::FromVariant;
 use iroha_schema::IntoSchema;
 use parity_scale_codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
 
-use crate::ffi;
+use crate::{conststr::ConstString, ffi};
+
+#[macro_export]
+/// Convenience macro to concisely construct a `[SocketAddr]`
+///
+/// # Examples
+/// ```
+/// # use iroha_primitives::socket_addr;
+///
+/// let localhost = socket_addr!(127,0,0,1;8080);
+/// let remote = socket_addr!("example.com";8080);
+/// ```
+macro_rules! socket_addr {
+    ($a:expr,$b:expr,$c:expr,$d:expr;$port:expr) => {
+        $crate::addr::SocketAddr::Ipv4($crate::addr::SocketAddrV4 {
+            ip: $crate::addr::Ipv4Addr::new([$a, $b, $c, $d]),
+            port: $port,
+        })
+    };
+    ($a:expr,$b:expr,$c:expr,$d:expr,$e:expr,$f:expr,$g:expr,$h:expr;$port:expr) => {
+        $crate::addr::SocketAddr::Ipv6($crate::addr::SocketAddrV6 {
+            ip: $crate::addr::Ipv6Addr::new([$a, $b, $c, $d, $e, $f, $g]),
+            port: $port,
+        })
+    };
+    ($host:expr;$port:expr) => {
+        $crate::addr::SocketAddr::Host($crate::addr::SocketAddrHost {
+            host: ($host).into(),
+            port: $port,
+        })
+    };
+}
+
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+#[derive(Debug, Clone, Display, Copy, PartialEq, Eq)]
+/// Error when parsing IP or socket address
+pub enum ParseError {
+    /// Not enough segments in IP address
+    NotEnoughSegments,
+    /// Too many segments in IP address
+    TooManySegments,
+    /// Failed to parse segment in IP address
+    InvalidSegment,
+    /// Failed to parse port number in socket address
+    InvalidPort,
+    /// Failed to find port number in socket address
+    NoPort,
+    /// Ipv6 address contains more than one '::' abbreviation
+    UnexpectedAbbreviation,
+}
 
 ffi::ffi_item! {
-    /// An Iroha-native version of `std::net::Ipv4Addr`, duplicated here to remain `no_std` compatible.
-    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, From, AsRef, IntoIterator, Decode, Encode, Deserialize, Serialize, IntoSchema)]
-    #[serde(transparent)]
+    /// An Iroha-native version of `std::net::Ipv4Addr`, duplicated here
+    /// to remain `no_std` compatible.
```

Comment:
But `SocketAddr` is available in [`core`}(https://doc.rust-lang.org/stable/core/net/enum.SocketAddr.html) and since `1.69` construction will be available in cons context.

Reply:
As far as I understand it's still not SCALE-compatible.

Reply:
By SCALE-compatible you mean that it doesn't implement `Encode/Decode`?